cudax/include/cuda/experimental/__memory_resource/memory_pool_base.cuh

```diff
@@ -0,0 +1,401 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__MEMORY_RESOURCE_MEMORY_POOL_BASE
+#define _CUDAX__MEMORY_RESOURCE_MEMORY_POOL_BASE
+
+#include <cuda/std/detail/__config>
+
+#include "cuda/std/__internal/namespaces.h"
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if _CCCL_CUDA_COMPILER(CLANG)
+#  include <cuda_runtime.h>
+#  include <cuda_runtime_api.h>
+#endif // _CCCL_CUDA_COMPILER(CLANG)
+
+#include <cuda/std/__cuda/api_wrapper.h>
+#include <cuda/std/span>
+
+#include <cuda/experimental/__stream/internal_streams.cuh>
+
+//! @file
+//! The \c __memory_pool_base class provides a wrapper around a `cudaMempool_t`.
+namespace cuda::experimental
+{
+
+enum class __memory_location_type
+{
+  __device,
+  __host,
+};
+
+//! @brief  Checks whether the current device supports \c cudaMallocAsync.
+//! @param __device_id The id of the device for which to query support.
+//! @throws cuda_error if \c cudaDeviceGetAttribute failed.
+//! @returns true if \c cudaDevAttrMemoryPoolsSupported is not zero.
+inline void __device_supports_stream_ordered_allocations(const int __device_id)
```

Comment:
I believe I came up with that name, but its bad. Because this does not retunrs anything. 

This should rather be prefixed with something like `__check` or `__verify`